client/src/nonblocking/rpc_client.rs

```diff
@@ -0,0 +1,5065 @@
+/// `client_error::Result` and [`RpcResult`] contain `ClientError` on error. In
+/// the case of `RpcResult`, the actual return value is in the
+/// [`value`][crate::rpc_response::Response::value] field, with RPC contextual
+/// information in the [`context`][crate::rpc_response::Response::context]
+/// field, so it is common for the value to be accessed with `?.value`, as in
+///
+/// ```
+/// # use solana_sdk::system_transaction;
+/// # use solana_client::rpc_client::RpcClient;
+/// # use solana_client::client_error::ClientError;
+/// # use solana_sdk::signature::{Keypair, Signer};
+/// # use solana_sdk::hash::Hash;
+/// # let rpc_client = RpcClient::new_mock("succeeds".to_string());
+/// # let key = Keypair::new();
+/// # let to = solana_sdk::pubkey::new_rand();
+/// # let lamports = 50;
+/// # let latest_blockhash = Hash::default();
+/// # let tx = system_transaction::transfer(&key, &to, lamports, latest_blockhash);
+/// let signature = rpc_client.send_transaction(&tx)?;
+/// let statuses = rpc_client.get_signature_statuses(&[signature])?.value;
+/// # Ok::<(), ClientError>(())
+/// ```
+///
+/// Requests may timeout, in which case they return a [`ClientError`] where the
+/// [`ClientErrorKind`] is [`ClientErrorKind::Reqwest`], and where the interior
+/// [`reqwest::Error`](crate::client_error::reqwest::Error)s
+/// [`is_timeout`](crate::client_error::reqwest::Error::is_timeout) method
+/// returns `true`. The default timeout is 30 seconds, and may be changed by
+/// calling an appropriate constructor with a `timeout` parameter.
+pub struct RpcClient {
+    sender: Box<dyn RpcSender + Send + Sync + 'static>,
+    config: RpcClientConfig,
+    node_version: RwLock<Option<semver::Version>>,
+}
+
+impl RpcClient {
+    /// Create an `RpcClient` from an [`RpcSender`] and an [`RpcClientConfig`].
+    ///
+    /// This is the basic constructor, allowing construction with any type of
+    /// `RpcSender`. Most applications should use one of the other constructors,
+    /// such as [`new`] and [`new_mock`], which create an `RpcClient`
+    /// encapsulating an [`HttpSender`] and [`MockSender`] respectively.
+    pub(crate) fn new_sender<T: RpcSender + Send + Sync + 'static>(
+        sender: T,
+        config: RpcClientConfig,
+    ) -> Self {
+        Self {
+            sender: Box::new(sender),
+            node_version: RwLock::new(None),
+            config,
+        }
+    }
+
+    /// Create an HTTP `RpcClient`.
+    ///
+    /// The URL is an HTTP URL, usually for port 8899, as in
+    /// "http://localhost:8899".
+    ///
+    /// The client has a default timeout of 30 seconds, and a default [commitment
+    /// level][cl] of [`Finalized`](CommitmentLevel::Finalized).
+    ///
+    /// [cl]: https://docs.solana.com/developing/clients/jsonrpc-api#configuring-state-commitment
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use solana_client::rpc_client::RpcClient;
```

Comment:
All the `fn` examples are unmodified from the original `RpcClient`.  Adapting them to `async` can be follow-up work